### PR TITLE
Add searching by array index

### PR DESCRIPTION
**Pull request by dan4thewin**
This PR includes full unit test line and branch coverage, and full CBMC proof coverage.  The code also passes coverity static scans and MISRA checks without error.

### CODE REVIEW

File: source/include/core_json.h [hunk 1]
```diff
@@ -84,18 +84,21 @@ JSONStatus_t JSON_Validate( const char * buf,
...
  *
- * The JSON document must contain an object (e.g., <code>{"key":"value"}</code>).
- * Any value may also be an object and so forth to a maximum depth.  A search
- * may descend through nested objects when the queryKey contains matching
- * key strings joined by a separator.
+ * Any value may also be an object or an array to a maximum depth.  A search
+ * may descend through nested objects or arrays when the queryKey contains matching
+ * key strings or array indexes joined by a separator.
  *
  * For example, if buf contains <code>{"foo":"abc","bar":{"foo":"xyz"}}</code>,
```
Comment (aggarw13) [inline source/include/core_json.h:94]: Better to use a proper phrase for the buffer or refer to the input parameter as `@p buf`
```suggestion
 * For example, if the provided buffer contains <code>{"foo":"abc","bar":{"foo":"xyz"}}</code>,
```

File: source/include/core_json.h [hunk 2]
```diff
@@ -118,9 +120,8 @@ JSONStatus_t JSON_Validate( const char * buf,
  *
  * @return #JSONSuccess if the queryKey is found and the value output;
  * #JSONNullParameter if any pointer parameters are NULL;
- * #JSONBadParameter if the queryKey is empty, or any subpart is empty, or max is 0;
- * #JSONIllegalDocument if the buffer contents are NOT valid JSON;
- * #JSONMaxDepthExceeded if object and array nesting exceeds a threshold;
+ * #JSONBadParameter if the queryKey is empty, or any subpart is empty, or max is 0,
```
Comment (aggarw13) [inline source/include/core_json.h:123]: What does "subpart is empty" mean? Can you update the wording to be clearer?

File: source/include/core_json.h [hunk 3]
```diff
@@ -106,7 +109,6 @@ JSONStatus_t JSON_Validate( const char * buf,
  * @param[in] max  size of the buffer.
  * @param[in] queryKey  The key to search for.
```
Comment (aggarw13) [inline source/include/core_json.h:110]: Should the parameter be renamed to `queryKeyOrIndex` now that this function supports both keys and array indices?

File: source/include/core_json.h [hunk 4]
```diff
@@ -106,7 +109,6 @@ JSONStatus_t JSON_Validate( const char * buf,
...
  * @param[in] queryKeyLength  Length of the key.
- * @param[in] separator  A character between a key and a sub-key in queryKey.
```
Comment (aggarw13) [inline source/include/core_json.h:109]: Why was this removed?

File: source/include/core_json.h [hunk 5]
```diff
@@ -118,9 +120,8 @@ JSONStatus_t JSON_Validate( const char * buf,
  *
  * @return #JSONSuccess if the queryKey is found and the value output;
  * #JSONNullParameter if any pointer parameters are NULL;
- * #JSONBadParameter if the queryKey is empty, or any subpart is empty, or max is 0;
- * #JSONIllegalDocument if the buffer contents are NOT valid JSON;
- * #JSONMaxDepthExceeded if object and array nesting exceeds a threshold;
+ * #JSONBadParameter if the queryKey is empty, or any subpart is empty, or max is 0,
+ * or an index is too large to convert;
```
Comment (aggarw13) [inline source/include/core_json.h:124]: The "index is too large to convert" phrase is unclear to me. What is the input index value converted to?

File: source/include/core_json.h [hunk 6]
```diff
@@ -106,7 +109,6 @@ JSONStatus_t JSON_Validate( const char * buf,
  * @param[in] max  size of the buffer.
```
Comment (aggarw13) [inline source/include/core_json.h:109]: Why isn't `bufSize` or `bufferSize` used as the parameter name instead of `max`? The use of `max` suggested **max depth** to be instead of buffer size when I saw the name..

File: source/include/core_json.h [hunk 7]
```diff
@@ -118,9 +120,8 @@ JSONStatus_t JSON_Validate( const char * buf,
...
  * @return #JSONSuccess if the queryKey is found and the value output;
  * #JSONNullParameter if any pointer parameters are NULL;
- * #JSONBadParameter if the queryKey is empty, or any subpart is empty, or max is 0;
- * #JSONIllegalDocument if the buffer contents are NOT valid JSON;
- * #JSONMaxDepthExceeded if object and array nesting exceeds a threshold;
+ * #JSONBadParameter if the queryKey is empty, or any subpart is empty, or max is 0,
+ * or an index is too large to convert;
  * #JSONNotFound if the queryKey is NOT found.
  *
  * <b>Example</b>
```
Comment (aggarw13) [inline source/include/core_json.h:127]: The example needs to be updated to not provide the separator character. 
Also, would it be helpful to provide 2 different examples, one for a JSON object and another for JSON array in the code snippet?

File: source/include/core_json.h [hunk 8]
```diff
@@ -155,15 +156,17 @@ JSONStatus_t JSON_Validate( const char * buf,
  *         value[ valueLength ] = save;
  *     }
  * @endcode
+ *
+ * @note The maximum index value is ~2 billion ( 2^31 - 9 ).
  */
 /* @[declare_json_search] */
 JSONStatus_t JSON_Search( char * buf,
```
Comment (aggarw13) [inline source/include/core_json.h:163]: Is there a reason why the buffer isn't of a `const char *` type considering that it is an input parameter?

File: source/include/core_json.h [hunk 9]
```diff
@@ -155,15 +156,17 @@ JSONStatus_t JSON_Validate( const char * buf,
  *         value[ valueLength ] = save;
  *     }
  * @endcode
+ *
+ * @note The maximum index value is ~2 billion ( 2^31 - 9 ).
```
Comment (aggarw13) [inline source/include/core_json.h:160]: What is the reason for this max value? May be, an explanatory comment could help

File: source/include/core_json.h [hunk 10]
```diff
@@ -155,15 +156,17 @@ JSONStatus_t JSON_Validate( const char * buf,
...
 /* @[declare_json_search] */
 JSONStatus_t JSON_Search( char * buf,
                           size_t max,
                           const char * queryKey,
                           size_t queryKeyLength,
-                          char separator,
                           char ** outValue,
                           size_t * outValueLength );
 /* @[declare_json_search] */
+#define MAX_INDEX_VALUE    ( 0x7FFFFFF7 )   /* 2^31 - 9 */
```
Comment (aggarw13) [inline source/include/core_json.h:170]: The max index value has to be passed as an ASCII string to the function. Does providing the max index value as an integer value provide use for the customer? (Should it be instead provided as an ASCII string?)

File: source/core_json.c [hunk 11]
```diff
@@ -1346,39 +1582,18 @@ JSONStatus_t JSON_Search( char * buf,
...
 
     if( ret == JSONSuccess )
     {
-        *outValue = p;
-        *outValueLength = tmp;
+        /* String values and collections include their surrounding
+         * demarcation.  If the value is a string, strip the quotes. */
+        if( *outValue[ 0 ] == '"' )
+        {
+            ( *outValue )++;
```
Comment (aggarw13) [inline source/core_json.c:1594]: Shouldn't this be an increment of the pointer address value instead of the pointed _to_ value?
```suggestion
            outValue++;
```

File: source/core_json.c [hunk 12]
```diff
@@ -1283,33 +1348,210 @@ static JSONStatus_t search( char * buf,
...
+static JSONStatus_t multiSearch( char * buf,
+                                 size_t max,
+                                 const char * queryKey,
+                                 size_t queryKeyLength,
+                                 char ** outValue,
+                                 size_t * outValueLength )
+{
+    JSONStatus_t ret = JSONSuccess;
+    size_t i = 0, start = 0, keyLength = 0;
+    char * p = buf;
```
Comment (aggarw13) [inline source/core_json.c:1495]: We probably should use more expressive variable names that convey the meaning of the variable. Referring to `i` and `p` variable names

File: source/core_json.c [hunk 13]
```diff
@@ -1283,33 +1348,210 @@ static JSONStatus_t search( char * buf,
...
+            }
+
+            i++;
+
+            found = arraySearch( p, tmp, ( uint32_t ) queryIndex, &p, &tmp );
         }
         else
         {
-            ret = JSONIllegalDocument;
+            start = i;
```
Comment (aggarw13) [inline source/core_json.c:1526]: The scope of `keyLength` variable can be reduced to this block
```suggestion
            size_t keyLength = 0U;
            start = i;
            
```

File: source/core_json.c [hunk 14]
```diff
@@ -1138,6 +1167,55 @@ JSONStatus_t JSON_Validate( const char * buf,
...
+ * @param[out] value  A pointer to receive the index of the value.
+ * @param[out] valueLength  A pointer to receive the length of the value.
+ *
+ * @return true if a value was present;
+ * false otherwise.
+ */
+static bool_ nextValue( const char * buf,
+                        size_t * start,
+                        size_t max,
+                        size_t * value,
```
Comment (aggarw13) [inline source/core_json.c:1188]: Suggestion for improving parameter name. `value` to be suggest the _value_ at the index in the buffer instead of the index integer value itself
```suggestion
                        size_t * index,
```

File: source/core_json.c [hunk 15]
```diff
@@ -1138,6 +1167,55 @@ JSONStatus_t JSON_Validate( const char * buf,
...
+    assert( ( value != NULL ) && ( valueLength != NULL ) );
+
+    i = *start;
+    valueStart = i;
+
+    if( ( skipAnyScalar( buf, &i, max ) == true ) ||
+        ( skipCollection( buf, &i, max ) == JSONSuccess ) )
+    {
+        *value = valueStart;
+        *valueLength = i - valueStart;
```
Comment (aggarw13) [inline source/core_json.c:1204]: Does this mean that the input `*start` value should already be representing the next value's index?

File: source/core_json.c [hunk 16]
```diff
@@ -1138,6 +1167,55 @@ JSONStatus_t JSON_Validate( const char * buf,
...
+ * @param[in,out] start  The index at which to begin.
+ * @param[in] max  The size of the buffer.
+ * @param[out] value  A pointer to receive the index of the value.
+ * @param[out] valueLength  A pointer to receive the length of the value.
+ *
+ * @return true if a value was present;
+ * false otherwise.
+ */
+static bool_ nextValue( const char * buf,
+                        size_t * start,
```
Comment (aggarw13) [inline source/core_json.c:1186]: Better naming suggestion
```suggestion
                        size_t * startIndex,
```

File: source/core_json.c [hunk 17]
```diff
@@ -1283,33 +1348,210 @@ static JSONStatus_t search( char * buf,
...
+ */
+static bool_ arraySearch( char * buf,
+                          size_t max,
+                          uint32_t queryIndex,
+                          char ** outValue,
+                          size_t * outValueLength )
+{
+    bool_ ret = false;
+    size_t i = 0, value, valueLength;
+    uint32_t j = 0;
```
Comment (aggarw13) [inline source/core_json.c:1384]: `j` is not easy to follow...
```suggestion
    uint32_t arrayIndex = 0;
```

File: source/core_json.c [hunk 18]
```diff
@@ -1283,33 +1348,210 @@ static JSONStatus_t search( char * buf,
...
+    i = *start;
+
+    while( ( i < max ) &&
+           !isSeparator_( buf[ i ] ) &&
+           !isSquareOpen_( buf[ i ] ) )
+    {
+        i++;
+    }
+
+    if( i > *start )
```
Comment (aggarw13) [inline source/core_json.c:1459]: Does this mean that if the input `*start` index value represents a query separator or an open square bracket, this function would return false?
Should the documentation of the function mention that the input index shouldn't point to either of the 2 locations within a _valid_ query string?

File: source/core_json.c [hunk 19]
```diff
@@ -1283,33 +1348,210 @@ static JSONStatus_t search( char * buf,
...
+ * @param[in] buf  The buffer to parse.
+ * @param[in,out] start  The index at which to begin.
+ * @param[in] max  The size of the buffer.
+ *
+ * @return true if a valid string was present;
+ * false otherwise.
+ */
+#define JSON_QUERY_KEY_SEPARATOR    '.'
+#define isSeparator_( x )    ( ( x ) == JSON_QUERY_KEY_SEPARATOR )
+static bool_ skipQueryKey( const char * buf,
```
Comment (aggarw13) [inline source/core_json.c:1439]: `advanceToQueryKeySubpart` is an easier name to understand the purpose of this function (even though it is wordy). **skipQueryKey** is ambiguous (imo).
```suggestion
static bool_ advanceToQueryKeySubpart( const char * buf,
```

File: source/core_json.c [hunk 20]
```diff
@@ -1283,33 +1348,210 @@ static JSONStatus_t search( char * buf,
...
+
+    return ret;
+}
+
+/**
+ * @brief Advance buffer index beyond a query key subpart.
+ *
+ * @param[in] buf  The buffer to parse.
+ * @param[in,out] start  The index at which to begin.
+ * @param[in] max  The size of the buffer.
```
Comment (aggarw13) [inline source/core_json.c:1432]: Missing `@param[out] outLength`

File: source/core_json.c [hunk 21]
```diff
@@ -1233,21 +1301,19 @@ static bool_ nextKeyValuePair( const char * buf,
...
  *
  * @note Parsing stops upon finding a match.
  */
-static JSONStatus_t search( char * buf,
-                            size_t max,
-                            const char * queryKey,
-                            size_t queryKeyLength,
-                            char ** outValue,
-                            size_t * outValueLength )
+static bool_ objectSearch( char * buf,
```
Comment (aggarw13) [inline source/core_json.c:1309]: Should the `buf` parameter be of `const char *` type?

File: source/include/core_json.h (hunk 1 above)
Comment (dan4thewin) [reply source/include/core_json.h:94]: changed

File: source/include/core_json.h (hunk 2 above)
Comment (dan4thewin) [reply source/include/core_json.h:123]: updated

File: source/include/core_json.h (hunk 3 above)
Comment (dan4thewin) [reply source/include/core_json.h:110]: I'll change it to be query, which is generic.

File: source/include/core_json.h (hunk 4 above)
Comment (dan4thewin) [reply source/include/core_json.h:109]: The array support introduces square brackets to surround an index in a query.  Rather than add code to permit the brackets to be user specified and code to ensure no collisions among separators, I decided that a user-specified separator was a misfeature and removed it.

File: source/include/core_json.h (hunk 5 above)
Comment (dan4thewin) [reply source/include/core_json.h:124]: as int32_t

File: source/include/core_json.h (hunk 6 above)
Comment (dan4thewin) [reply source/include/core_json.h:109]: Well, it made sense to me and the previous reviewers.  It's part of almost every function.

File: source/include/core_json.h (hunk 7 above)
Comment (dan4thewin) [reply source/include/core_json.h:127]: fixed
More is usually better, but with our time window, I'll demur.

File: source/include/core_json.h (hunk 8 above)
Comment (dan4thewin) [reply source/include/core_json.h:163]: Yes, since the outputs alias the input, it makes a mess for the caller w.r.t. warnings and MISRA.

File: source/include/core_json.h (hunk 9 above)
Comment (dan4thewin) [reply source/include/core_json.h:160]: The algorithm is the same as you'd find in most any base 10 conversion.  Given our time constraints, I don't want to play the pundit.

File: source/include/core_json.h (hunk 10 above)
Comment (dan4thewin) [reply source/include/core_json.h:170]: Really, no one should be iterating over 2 billion elements into a JSON array.  I originally put in the calculation seen in the comment, but struggle with the MISRA complaints that ensued.

File: source/core_json.c (hunk 11 above)
Comment (dan4thewin) [reply source/core_json.c:1594]: outValue is a char **.  If this were wrong, the unit tests would catch it.

File: source/core_json.c (hunk 12 above)
Comment (dan4thewin) [reply source/core_json.c:1495]: These are idiomatic usages, and passed muster with previous reviewers.  `i` is used in the same fashion in almost every function.  We can discuss whether idioms are a good or bad thing, but I'd rather not do so in the context of this PR, given our time considerations.

File: source/core_json.c (hunk 13 above)
Comment (dan4thewin) [reply source/core_json.c:1526]: changed

File: source/core_json.c (hunk 14 above)
Comment (dan4thewin) [reply source/core_json.c:1188]: I think `index` is too generic here.  `value` is described as an index in the doc comment immediately above the function.  I'm going to leave it as is.

File: source/core_json.c (hunk 15 above)
Comment (dan4thewin) [reply source/core_json.c:1204]: yes

File: source/core_json.c (hunk 16 above)
Comment (dan4thewin) [reply source/core_json.c:1186]: I think `start` is fine.  It's used in the majority of functions in the same fashion.

File: source/core_json.c (hunk 17 above)
Comment (dan4thewin) [reply source/core_json.c:1384]: Changed to `currentIndex`.

File: source/core_json.c (hunk 18 above)
Comment (dan4thewin) [reply source/core_json.c:1459]: Yes.  Added a comment.

File: source/core_json.c (hunk 19 above)
Comment (dan4thewin) [reply source/core_json.c:1439]: changed to `skipQueryPart`

File: source/core_json.c (hunk 20 above)
Comment (dan4thewin) [reply source/core_json.c:1432]: added

File: source/core_json.c (hunk 21 above)
Comment (dan4thewin) [reply source/core_json.c:1309]: No, because of the aliasing I mentioned above.

File: source/core_json.c (hunk 15 above)
Comment (aggarw13) [reply source/core_json.c:1204]: It would be good to mention that as a _requirement_ in the function documentation.

File: source/include/core_json.h (hunk 6 above)
Comment (aggarw13) [reply source/include/core_json.h:109]: It wouldn't hurt to update just the public-facing functions' parameters to make it easier for the customer to understand. I am fine with keeping `max` as-is within the private functions in the implementation

File: source/include/core_json.h [hunk 22]
```diff
@@ -84,18 +84,21 @@ JSONStatus_t JSON_Validate( const char * buf,
...
+ * key strings or array indexes joined by a separator.
  *
- * For example, if buf contains <code>{"foo":"abc","bar":{"foo":"xyz"}}</code>,
+ * For example, if the provided buffer contains <code>{"foo":"abc","bar":{"foo":"xyz"}}</code>,
  * then a search for 'foo' would output <code>abc</code>, 'bar' would output
  * <code>{"foo":"xyz"}</code>, and a search for 'bar.foo' would output
- * <code>xyz</code> (given separator is specified as '.').
+ * <code>xyz</code>.
+ *
+ * If buf contains <code>[123,456,{"foo":"abc","bar":[88,99]}]</code>,
```
Comment (aggarw13) [inline source/include/core_json.h:99]:
```suggestion
 * If the provided buffer contains <code>[123,456,{"foo":"abc","bar":[88,99]}]</code>,
```

File: source/include/core_json.h (hunk 10 above)
Comment (aggarw13) [reply source/include/core_json.h:176]: What is the utility of exposing this macro in the API if the user _should_ not be using it. Can we just move it to within the source file?

File: source/include/core_json.h [hunk 23]
```diff
@@ -116,21 +118,20 @@ JSONStatus_t JSON_Validate( const char * buf,
  * @note JSON_Search() performs validation, but stops upon finding a matching
  * key and its value. To validate the entire JSON document, use JSON_Validate().
  *
- * @return #JSONSuccess if the queryKey is found and the value output;
+ * @return #JSONSuccess if the query is matched and the value output;
```
Comment (aggarw13) [inline source/include/core_json.h:121]: "and the value output" seems incomplete

File: test/unit-test/core_json_utest.c [hunk 24]
```diff
@@ -1589,4 +1658,13 @@ void test_JSON_unreached( void )
     buf[ 0 ] = 'x';
     start = 0;
     TEST_ASSERT_EQUAL( false, skipEscape( buf, &start, sizeof( buf ) ) );
+
+    /* set output value to -1 when integer conversion exceeds max */
+    {
+#define TOO_BIG    "100000000000"
```
Comment (aggarw13) [inline test/unit-test/core_json_utest.c:1664]: Should you also test this value with `JSON_Search`?

File: lexicon.txt [hunk 25]
```diff
@@ -51,12 +51,13 @@ nb
 nextkeyvaluepair
 noninfringement
 nul
```
Comment (sarenameas) [inline lexicon.txt:53]: Can we delete this misspelled word? It looks like its not existent in the library.

File: source/include/core_json.h [hunk 26]
```diff
@@ -116,21 +118,20 @@ JSONStatus_t JSON_Validate( const char * buf,
...
  * #JSONNullParameter if any pointer parameters are NULL;
- * #JSONBadParameter if the queryKey is empty, or any subpart is empty, or max is 0;
- * #JSONIllegalDocument if the buffer contents are NOT valid JSON;
- * #JSONMaxDepthExceeded if object and array nesting exceeds a threshold;
- * #JSONNotFound if the queryKey is NOT found.
+ * #JSONBadParameter if the query is empty, or the portion after a separator is empty,
+ * or max is 0, or an index is too large to convert;
```
Comment (aggarw13) [inline source/include/core_json.h:124]: Improve clarity about cause of error from passed index in query string
```suggestion
 * or max is 0, or an index is too large to convert to an integer value;
```

File: lexicon.txt (hunk 25 above)
Comment (dan4thewin) [reply lexicon.txt:53]: NUL is the ASCII name for '\0'.  It is in the comment for skipEscape().

File: source/include/core_json.h (hunk 26 above)
Comment (dan4thewin) [reply source/include/core_json.h:124]: changed

File: test/unit-test/core_json_utest.c (hunk 24 above)
Comment (dan4thewin) [reply test/unit-test/core_json_utest.c:1664]: I think the tests are adequate.  Branch coverage is 100%.

File: source/include/core_json.h (hunk 23 above)
Comment (dan4thewin) [reply source/include/core_json.h:121]: Here, output is a verb.

File: source/include/core_json.h (hunk 22 above)
Comment (dan4thewin) [reply source/include/core_json.h:99]: changed

File: source/core_json.c (hunk 15 above)
Comment (dan4thewin) [reply source/core_json.c:1204]: added a comment

File: source/include/core_json.h (hunk 10 above)
Comment (dan4thewin) [reply source/include/core_json.h:176]: I reference the macro in tests.

File: source/include/core_json.h (hunk 6 above)
Comment (dan4thewin) [reply source/include/core_json.h:109]: Changing the name in the declaration and not in the implementation throws a MISRA 8.3 error.  I'd rather have consistency in core_json.c.

**Review by aggarw13 — APPROVED** [review]

**Review by cobusve — APPROVED** [review]
